presto-main/src/main/java/com/facebook/presto/type/MapType.java

```diff
@@ -69,6 +72,107 @@ public Type getValueType()
     }
 
     @Override
+    public boolean isComparable()
+    {
+        return valueType.isComparable();
+    }
+
+    @Override
+    public int hash(Block block, int position)
+    {
+        Slice mapSlice = getSlice(block, position);
+        Block mapBlock = readStructuralBlock(mapSlice);
+        int result = 0;
+
+        for (int i = 0; i < mapBlock.getPositionCount(); i += 2) {
+            result += hashPosition(keyType, mapBlock, position);
+            result += hashPosition(valueType, mapBlock, position + 1);
+        }
+        return result;
+    }
+
+    @Override
+    public boolean equalTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
+    {
+        Slice leftSlice = getSlice(leftBlock, leftPosition);
+        Slice rightSlice = getSlice(rightBlock, rightPosition);
+        Block leftMapBlock = readStructuralBlock(leftSlice);
+        Block rightMapBlock = readStructuralBlock(rightSlice);
+
+        if (leftMapBlock.getPositionCount() != rightMapBlock.getPositionCount()) {
+            return false;
+        }
+
+        Map<KeyWrapper, Integer> wrappedLeftMap = new HashMap<>();
+        for (int position = 0; position < leftMapBlock.getPositionCount(); position += 2) {
+            wrappedLeftMap.put(new KeyWrapper(keyType, leftMapBlock, position), position + 1);
+        }
+
+        Map<KeyWrapper, Integer> wrappedRightMap = new HashMap<>();
+        for (int position = 0; position < rightMapBlock.getPositionCount(); position += 2) {
+            wrappedRightMap.put(new KeyWrapper(keyType, rightMapBlock, position), position + 1);
+        }
+
+        for (Map.Entry<KeyWrapper, Integer> entry : wrappedRightMap.entrySet()) {
```

Comment:
There's one more optimization we can make here. You can fuse this loop with the one above, since you don't need to construct the `wrappedRightMap`. Just iterate over the positions in`rightMapBlock` directly instead of `wrappedRightMap.entrySet()`